Firestore/core/src/firebase/firestore/api/document_reference.mm

```diff
@@ -0,0 +1,227 @@
+      // If we're online and the document doesn't exist then we call the
+      // completion with a document with document.exists set to false. If we're
+      // offline however, we call the completion handler with an error. Two
+      // options: 1) Cache the negative response from the server so we can
+      // deliver that even when you're
+      //    offline.
+      // 2) Actually call the completion handler with an error if the document
+      // doesn't exist when
+      //    you are offline.
+      completion(nil, [NSError errorWithDomain:FIRFirestoreErrorDomain
+                                          code:FIRFirestoreErrorCodeUnavailable
+                                      userInfo:@{
+                                        NSLocalizedDescriptionKey :
+                                            @"Failed to get document because "
+                                            @"the client is offline.",
+                                      }]);
+    } else if (snapshot.exists && snapshot.metadata.fromCache &&
+               source == FIRFirestoreSourceServer) {
+      completion(
+          nil,
+          [NSError
+              errorWithDomain:FIRFirestoreErrorDomain
+                         code:FIRFirestoreErrorCodeUnavailable
+                     userInfo:@{
+                       NSLocalizedDescriptionKey :
+                           @"Failed to get document from server. (However, "
+                           @"this "
+                           @"document does exist in the local cache. Run again "
+                           @"without setting source to "
+                           @"FIRFirestoreSourceServer to "
+                           @"retrieve the cached document.)"
+                     }]);
+    } else {
+      completion(snapshot, nil);
+    }
+  };
+
+  listener_registration = AddSnapshotListener(listener, options);
+  dispatch_semaphore_signal(registered);
+}
+
+id<FIRListenerRegistration> DocumentReference::AddSnapshotListener(
+    FIRDocumentSnapshotBlock listener, FSTListenOptions* options) {
+  FIRFirestore* firestore = firestore_;
+  FSTQuery* query = [FSTQuery queryWithPath:key_.path()];
+  DocumentKey key = key_;
+
+  ViewSnapshotHandler snapshotHandler = [key, listener, firestore](
+                                            const StatusOr<ViewSnapshot>&
+                                                maybe_snapshot) {
+    if (!maybe_snapshot.ok()) {
+      listener(nil, MakeNSError(maybe_snapshot.status()));
+      return;
+    }
+
+    const ViewSnapshot& snapshot = maybe_snapshot.ValueOrDie();
+    HARD_ASSERT(snapshot.documents().count <= 1,
+                "Too many document returned on a document query");
+    FSTDocument* document = [snapshot.documents() documentForKey:key];
+
+    bool has_pending_writes =
+        document
+            ? snapshot.mutated_keys().contains(key)
+            : NO;  // We don't raise `has_pending_writes` for deleted documents.
```

Comment:
s/NO/false/g; s/YES/true/g throughout this file.

Reply:
Done.